torch/autograd/__init__.py

```diff
@@ -28,6 +28,12 @@ def _make_grads(outputs, grads):
                                    + str(grad.shape) + " and output["
                                    + str(outputs.index(out)) + "] has a shape of "
                                    + str(out.shape) + ".")
+            if (out.dtype.is_complex and not grad.dtype.is_complex) or (not out.dtype.is_complex and grad.dtype.is_complex):
+                raise RuntimeError("Mismatch in dtype: grad_output["
```

Comment:
nit: given that this is only for complex, we could be more precise here and replace `dtype` by stating that the user is mixing complex and real Tensors.

Reply:
I'm not sure I understand where this is used -- why is it okay for the dtypes to differ at all?

Reply:
yeah I am not sure either. I saw some test failures when I enabled it more generally for all dtypes:
```
======================================================================
ERROR [0.003s]: test_no_grad (_main_.TestNN)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_nn.py", line 629, in test_no_grad
    output.backward(torch.ones(1, 5, 10, 10))
  File "C:\Users\circleci\project\build\win_tmp\build\torch\tensor.py", line 198, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\autograd\__init__.py", line 100, in backward
    grad_tensors = _make_grads(tensors, grad_tensors)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\autograd\__init__.py", line 36, in _make_grads
    + str(out.dtype) + ".")
RuntimeError: Mismatch in dtype: grad_output[0] has a dtype of torch.float64 and output[0] has a dtype of torch.bfloat16.
```

Reply:
```
======================================================================
FAIL [0.002s]: test_vjp_err_check_strict (_main_.TestAutogradFunctional)
----------------------------------------------------------------------
RuntimeError: Mismatch in dtype: grad_output[0] has a dtype of torch.float64 and output[0] has a dtype of torch.float32.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test_autograd.py", line 4352, in test_vjp_err_check_strict
    res = autogradF.vjp(bar, inp, v, strict=True)
AssertionError: "The output of the user-provided function is independent of input 0" does not match "Mismatch in dtype: grad_output[0] has a dtype of torch.float64 and output[0] has a dtype of torch.float32."
```

Reply:
I guess for low precision training? forward could be done in float16 but backward in float32? Otherwise it prevent convergence?